PROPOSAL: use datetime64 and timedelta64 to eval

holymonson opened the pull request:
This PR aims to use datetime64 and timedelta64 in evaluation directly, instead of casting to integer and then casting back. 

Benefits:
1. `values` is saved in `ndarray[datetime64]`, casting to `int` to eval is non-intuitive. It seems back to the time before numpy 1.7 which implements `datetime64`.
2. `numpy.ufunc` can handle all the `dtype` change, e.g.  `datetime64 - datetime64 = timedelta64`. Additonally, nonsense operations are prohibited, e.g. `datetime64 + datetime64`.
3. numpy implements `datetime64('NaT')`, seems `values_mask` and `other_mask` are no longer needed for `nan`s.
4. `Series` can share the same call path with `DataFrame` in `Block`. Now `Series` is using `dispatch_to_index_op` to wrap as `Index`, but `Index` is 1d-array while `DataFrame` is 2d-array, so `DataFrame` can't use the same wrapper. `ndarray([Index])` is `dtype='O'` so can't be benefited from `numpy.ufunc`.

This commit has passed `tests/{internals,frame,dtypes}` , but not `tests/series`, because `dispatch_to_index_op` breaks it.

holymonson committed:
use datetime64 and timedelta64 to eval

jbrockmendel commented:
> `Series` can share the same call path with `DataFrame` in `Block`

We actually want to go in the other direction.  We want to make `DataFrame` dispatch to the `Series` (or `DatetimeIndex`/`TimedeltaIndex`) implementations.  Is there a viable way to do that here?  Maybe dispatch to `self._holder`?

jbrockmendel commented on pandas/core/common.py:
```diff
@@ -431,3 +431,11 @@ def _pipe(obj, func, *args, **kwargs):
+
+
+# TODO: np.percentile not support datetime64, should be fixed in upsteam.
+def percentile(values, q, **kw):
```
Probably belongs in `core.nanops`, definitely not `core.common`.

jbrockmendel commented on pandas/core/common.py:
```diff
@@ -431,3 +431,11 @@ def _pipe(obj, func, *args, **kwargs):
+# TODO: np.percentile not support datetime64, should be fixed in upsteam.
+def percentile(values, q, **kw):
+    if values.dtype.kind == 'M':
+        return np.percentile(values.view('i8'), q, **kw).astype(values.dtype)
```
Will this handle `iNaT` correctly?

holymonson replied:
Can handle `numpy.datetime64('NaT')`, but not `iNaT`.

jbrockmendel commented on pandas/core/internals/blocks.py:
```diff
@@ -2134,7 +2134,7 @@ def _na_value(self):
     @property
     def fill_value(self):
-        return tslibs.iNaT
+        return tslibs.NaT.asm8
```
`iNaT` is the standard usage for this.

holymonson replied:
`iNaT ` is an `int`, but here wants a `numpy.datetime64('NaT')`.

jreback reviewed (changes requested):
what problem are you trying to solve here?

jbrockmendel commented on pandas/tests/internals/test_internals.py:
```diff
@@ -1250,14 +1250,20 @@ class TestCanHoldElement(object):
-                (operator.pow, 'bool')}
+                (operator.pow, 'bool'),
+                (operator.pow, '<m8[ns]'),
+                (operator.pow, '<M8[ns]')}
```
These are good fixes.  Can you take it a step further and, instead of skipping, `pytest.raises(TypeError)` for the appropriate cases?

holymonson committed:
block: save .tz in DatetimeTZBlock, discard DatetimeIndex

1. `.tz` info save as `DatetimeTZBlock.tz`, no longer need a `DatetimeIndex`. Multi-columns can share a same `DatetimeTZBlock` if they have same `tz`. (NotImplemented)
2. `_try_coerce_args` is implemented in `DatetimeLikeBlockMixin`, those override in subclass are removed. So as `_try_coerce_result`. (Benefited from `numpy` handing data)
3. `values_mask` and `other_mask` are also removed. (`numpy` can handle `NaT`)

pep8speaks commented:
Hello @holymonson! Thanks for updating the PR.



 - In the file [`pandas/core/internals/blocks.py`](https://github.com/pandas-dev/pandas/blob/e52ed3a5395c1e1fd8eb13acb1ebced58285a608/pandas/core/internals/blocks.py), following are the PEP8 issues :

> [Line 2148:80](https://github.com/pandas-dev/pandas/blob/e52ed3a5395c1e1fd8eb13acb1ebced58285a608/pandas/core/internals/blocks.py#L2148): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 2149:9](https://github.com/pandas-dev/pandas/blob/e52ed3a5395c1e1fd8eb13acb1ebced58285a608/pandas/core/internals/blocks.py#L2149): [E112](https://duckduckgo.com/?q=pep8%20E112) expected an indented block

holymonson committed:
compat: add np_percentile_compat for datetime64

holymonson commented on pandas/core/internals/blocks.py:
```diff
@@ -2145,6 +2128,32 @@ def get_values(self, dtype=None):
+            # let higher levels handle
+            raise TypeError
+
+        return values, other
```
Seems `values` can be removed too. The other 2 `_try_coerce_args` overrides don't touch `values`, `_try_coerce_args` might rename to `_try_coerce_other`.

jbrockmendel commented:
> can only handle 1d-array

Can you expand on why this is a compelling reason?  Maybe there is some characteristic np.ufunc that I'm not considering?

> so `DataFrame` can't use the same wrapper

The idea is that `DataFrame` dispatches column-wise, like it already does in `DataFrame._combine_frame` and `DataFrame._compare_frame`

> Could someone show me the benefits of the Index-dispatch implement?

See #18824.  As you've noticed, there are many operations (particularly arithmetic and comparisons) where `DataFrame` behavior is incorrect, particularly for datetime64 and datetime64tz dtypes.  Until a few months ago, the same was true for Series.  We fixed the Series problems by having it dispatch to the Index subclasses, since the DatetimeIndex and TimedeltaIndex implementations are much more careful.  Now we're down to two implementations instead of three; we still should get it down to one.

holymonson committed:
add pytest.raises to test_binop_other

jbrockmendel commented on pandas/compat/numpy/__init__.py:
```diff
@@ -68,6 +68,22 @@ def np_array_datetime64_compat(arr, *args, **kwargs):
 
+# np.percentile should support datetime64 in upsteam. numpy/numpy#11620
+# TODO: change the version if the bug fixed
+if _nlv >= '99.99.99':
```
Is this just a placeholder for "distant future"?

jbrockmendel commented on pandas/compat/numpy/__init__.py:
```diff
@@ -68,6 +68,22 @@ def np_array_datetime64_compat(arr, *args, **kwargs):
+        provide compat for np.percentile supporting datetime64
+        '''
+        if values.dtype.kind == 'M':
+            result = np.percentile(values.view('i8'), q, **kw)
```
Will this handle iNaT correctly?

jbrockmendel commented on pandas/compat/numpy/__init__.py:
```diff
@@ -68,6 +68,22 @@ def np_array_datetime64_compat(arr, *args, **kwargs):
+if _nlv >= '99.99.99':
+    np_percentile_compat = np.percentile
+else:
+    def np_percentile_compat(values, q, **kw):
```
the iNat concern aside, this looks like a nice function.  But I don't see why it should be in compat.  Might make more sense in nanops?

jbrockmendel commented on pandas/compat/numpy/__init__.py:
```diff
@@ -68,6 +68,22 @@ def np_array_datetime64_compat(arr, *args, **kwargs):
+        '''
+        provide compat for np.percentile supporting datetime64
+        '''
+        if values.dtype.kind == 'M':
```
Does timedelta64 get handled correctly by numpy?

jbrockmendel commented on pandas/core/internals/blocks.py:
```diff
@@ -789,8 +789,7 @@ def replace(self, to_replace, value, inplace=False, filter=None,
         try:
-            values, _, to_replace, _ = self._try_coerce_args(self.values,
-                                                             to_replace)
+            values, to_replace = self._try_coerce_args(self.values, to_replace)
```
It's hard to tell, but a lot of this looks like it might be unrelated to the np_percentile_compat function.  Is that the case?  If so, that's great (simplifying try_coerce_args etc would be a big win), but should be done separately.

jbrockmendel commented:
@holymonson can you rebase?  Arithmetic ops have seen a lot of cleanup in the last couple months.

WillAyd commented:
Closing as stale. Ping if you'd like to pick this back up
